### ISSUE TITLE
namespacesRequired, withNamespaces confusing

### ISSUE THREAD

**Issue by osdiab** [isaachinman/next-i18next#253]
The documentation over which of these does what is pretty confusing at the moment and I'm not sure exactly what I need to do/don't really understand how this tool works, which is giving me cold feet to rely on it.

In [the simple example](https://github.com/isaachinman/next-i18next/blob/master/examples/simple/pages/index.js), `namespacesRequired` passes down both `'common'` and `'footer'`, while `withNamespaces` only passes `'common'`. What exactly is going on here? Why are they different?

If they are supposed to be the same, the fact I need to supply them on both places seems kinda redundant, but I also don't really parse why both are required according to [the README](https://github.com/isaachinman/next-i18next#4-declaring-namespace-dependencies), so maybe I'm missing something.

It seems based on [this PR comment that took me a while to find](https://github.com/isaachinman/next-i18next/issues/54#issuecomment-447789150) that the page's required namespaces are not being inferred from the page's components' required namespaces, so I assume that's the difference - but then I'm confused exactly what `withNamespaces` at the component level does. Is it that `namespacesRequired` on a page tells SSR what to bundle together with the page, and that `withNamespaces` will load those either from the bundle if it was passed along, or issue a network request if it isn't? Wild guess of mine.

Basically I want to know what each field/component/HOC does and how they behave with different contexts. Would love some clarification, esp in the form of clearer docs!

**Comment by osdiab** [comment]
After struggling most of the day trying to get everything to work I think I understand better. The messages for a namespace are only made available to a component if `withNamespaces` specifies it, and the component can only access it if the page's `namespacesRequired` provides that namespace. If that's a correct understanding perhaps I can add/rewrite something in the README to clear that up.

**Comment by isaachinman** [comment]
First, let me apologise if the documentation was unclear, and that you had to work this out on your own.

> The messages for a namespace are only made available to a component if withNamespaces specifies it, and the component can only access it if the page's namespacesRequired provides that namespace.

That is indeed the correct understanding. In fact, `withNamespaces` is a direct copy from `react-i18next`, and simply provides namespaces via `t` (translate) function at the _component level_. The `namespacesRequired` prop from the _page level_ is a `next-i18next` thing, and tells our server what namespaces to include in the initial data when rendering that page out.

It's unfortunate that we've had to resort to devs managing `namespacesRequired` required themselves - my initial idea was to do tree traversal and collect all nested `withNamespaces` values. But it turns out using `react-tree-walker` or similar approaches is a big hit for performance because you essentially have to render the whole tree twice. Moreover, libraries like Apollo already do tree traversal unashamedly, so that'd mean three full renders for users of Apollo in conjunction with `next-i18next`, and so on.

Here's our current note in the docs:

> Note: withNamespaces provides namespaces to the component that it wraps. However, namespacesRequired provides the total available namespaces to the entire React tree and belongs on the page level.

How do you feel we can improve this, to save future users' time?

**Comment by isaachinman** [comment]
@osdiab Any update?

**Comment by osdiab** [comment]
Sorry for not responding!

I think there's a couple things worth mentioning:

1. `withNamespaces` comes from `reacti-18next`. I came into this library having never used either, so I think it makes sense to be aware of it if you were migrating from `react-i18next`, but for me that wasn't the case.
1. Because I don't know how `react-i18next` provides those namespaces, I don't actually know "what" is providing namespaces to the component, or how it does that.

So I think to make it clearer, I would do two things:

#### Section 2
Re: filesystem organization, describe what namespaces are, roughly, or link to relevant part of `i18next` that describes it. And include other namespaces in the example hierarchy than the `common` one.

```
As in `i18next`, translations are grouped into "namespaces" whose names are defined by their filenames in the locales directory.
```

#### Section 4
Then, with that context, this can be edited like so, to give context to the `withNamespaces` and `namespacesRequired` fields:

```
### 4. Declaring namespace dependencies

As in `react-i18next`, any React component that wishes to use translations must use the `withNamespaces` HOC to declare which namespaces it depends on. Alternatively, you can use `Trans` component instead of `withNamespaces` if desired. **INSERT AN EXAMPLE HERE** (I actually don't know how to properly use that to provide a namespace; is this actually true?)

Unlike `react-i18next`, however, each NextJS page may also declare which namespaces will be made available to the client. By default, `next-i18next` will send _all your namespaces_ down to the client on each initial request. This can be an appropriate approach for smaller apps with less content, but a lot of apps will benefit from splitting namespaces based on route.

To do that, you need to return a `namespacesRequired` array via `getInitialProps` on your page-level component. You can see this approach in [examples/simple/pages/index.js](./examples/simple/pages/index.js).

Note: `next-i18next` at this time does not analyze the React tree to determine what namespaces are required for a given page. That means that if a component uses the `withNamespaces` HOC to declare a namespace as a dependency, but the page's `namespacesRequired` field is present and does not make that namespace available, your component will be unable to properly display the desired translations. Therefore, if you specify the `namespacesRequired` field in a NextJS page, ensure that it includes all namespaces required by the components within that page.
```

**Comment by osdiab** [comment]
I can make a PR for the above if you'd like
